lib/logging.py

```diff
@@ -21,20 +28,32 @@ def logger() -> logging.Logger:
         return SecurityLog.seclogger
 
     @staticmethod
-    def logevent(request: HttpRequest, type: str, message: str) -> None:
-        # Unit tests do not have user defined in request object.
-        if request and hasattr(request, 'user'):
-            user=request.user
-            ip=request.META.get('REMOTE_ADDR')
-        else:
-            user='?'
-            ip='?'
-        SecurityLog.logger().info("({}/{}): {}: {}".format(
-            user,
-            ip,
-            type,
-            message
-        ))
+    def logevent(
+            request: HttpRequest,
+            type: str,
+            message: str,
+            severity: int = 5,
+            ) -> None:
+        extensions = {}
+        # Unit tests may not have user or source address defined.
+        if request:
+            if hasattr(request, 'user'):
```

Comment:
Unit tests do not seem to have user attribute set in HttpRequest, and therefore fail without hasattr check. I don't think any of the security log messages are reachable without logged in user, so couldn't test how it would behave with AnonymousUser.

I now also noticed that the change to failed login caused unit tests to fail because they do not have 'username' key defined, so did one more small fix to that.

Reply:
Unit tests may log in with the test client and simulate an authenticated user.

https://docs.djangoproject.com/en/3.2/topics/testing/tools/#django.test.Client.login